### PR TITLE
Speed up query by using only one indexed column in sql where clause

### PR DESCRIPTION
**Pull request by csthomas**
### Summary of Changes

* Usual mysql databese can only use one index in WHERE clause or use union.
* Newer version can merge indexes to speed up.

Current query contains `OR parent_id = 0` which complicate query optimisation for database.
This PR replace above to use only `name` column.

### Testing Instructions - Old
Unit tests pass. 
Code review.

On installed joomla with sample data, mysql may not use indexes because there is too less rows.

### Testing Instructions - New
Install joomla without patch, turn off cache and go to front page.
Joomla works OK without any errors. 

Apply patch and do the same.
Joomla works OK without any errors.

### Documentation Changes Required
N/A

### CODE REVIEW

File: libraries/joomla/access/access.php [hunk 1]
```diff
@@ -403,7 +403,7 @@ protected static function preloadComponents()
 		!JDEBUG ?: JProfiler::getInstance('Application')->mark('Before JAccess::preloadComponents (all components)');
 
 		// Add root to asset names list.
-		$components = array();
+		$components = array('root.1');
```
Comment (andrepereiradasilva) [inline libraries/joomla/access/access.php]: Is the root asset always `root.1`? And even so should i tbe hardcoded?
The `parent_id = 0` allows to not hardcode the root asset name here.

Comment (csthomas) [inline libraries/joomla/access/access.php]: I have not think about it too much. I will check it.

It was ok before your PR https://github.com/joomla/joomla-cms/commit/a6a0a96e3868ee4ccf3979206c99f4a29a0af3a4#diff-bbdd2441fb63d5dea47802e51fc9df75L364

Comment (csthomas) [inline libraries/joomla/access/access.php]: 'root.1' exists in the same file two times more. Maybe joomla can use `parent_id = 0` or 'root.1' depends which is more effective.

Comment (csthomas) [inline libraries/joomla/access/access.php]: If 'root.1' should not be use then I can redesign query and use UNION instead OR in SELECT query.

Comment (mbabker) [inline libraries/joomla/access/access.php]: `root.1` will always exist and will always be the root unless your assets table is FUBAR.  What's the intent here?

Comment (csthomas) [inline libraries/joomla/access/access.php]: I want to replace `asset_name IN (...) OR parent_id = 0` to simpler 'asset_name IN ('root.id', ...)'.
This way mysql does not have to scan table but only index.

If you talk about whole PR then if $components is empty then without this PR sql generates errors.

Comment (mbabker) [inline libraries/joomla/access/access.php]: Then `root.1` will always exist.

**Comment by csthomas** [comment]
My explanation for mariadb is a little outdated (because mariadb can merge indexes) but new query is better that older.

`EXPLAIN SELECT id, name, rules, parent_id FROM #__assets WHERE name IN ('root.1', 'com_users')`
```php
Array(
            [id] => 1
            [select_type] => SIMPLE
            [table] => j37_assets
            [partitions] => 
            [type] => range
            [possible_keys] => idx_asset_name
            [key] => idx_asset_name
            [key_len] => 202
            [ref] => 
            [rows] => 2
            [filtered] => 100.00
            [Extra] => Using index condition
        )
```

And older one:
`EXPLAIN SELECT id, name, rules, parent_id FROM #__assets WHERE name IN ('com_users') or parent_id=0`
```php
Array(
            [id] => 1
            [select_type] => SIMPLE
            [table] => j37_assets
            [partitions] => 
            [type] => index_merge
            [possible_keys] => idx_asset_name,idx_parent_id
            [key] => idx_asset_name,idx_parent_id
            [key_len] => 202,4
            [ref] => 
            [rows] => 2
            [filtered] => 100.00
            [Extra] => Using union(idx_asset_name,idx_parent_id); Using where
        )
```

File: libraries/joomla/access/access.php (hunk 1 above)
Comment (andrepereiradasilva) [inline libraries/joomla/access/access.php]: as michael comments, ok, then no issue

**Review by andrepereiradasilva — APPROVED** [review]

**Review by mbabker — APPROVED** [review]

**Comment by csthomas** [comment]
This PR has 2 approves and it should be enough as test based on code review.
RTC?

**Comment by franz-wohlkoenig** [comment]
RTC after two Approves.

**Comment by rdeutz** [comment]
approval is not tested, at least one test should be done

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on b61cfa960588ce6f7ebf009882dcd8b4fce59ad7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13528">issues.joomla.org/tracker/joomla-cms/13528</a>.</sub>

**Pull request merged**